File: Binary_tree_general/leetcode/230_Kth_Smallest_Element_in_a_BST/230_Kth_Smallest_Element_in_a_BST.c
#include <stdio.h>
#include <stdlib.h>

// 이진 트리 노드 구조체 정의
struct TreeNode {
    int val;
    struct TreeNode* left;
    struct TreeNode* right;
};

// 전역 변수 (상태 관리)
int count;    // 방문한 노드 개수
int result;   // k번째 노드의 값
int found;    // k번째 노드를 찾았는지 여부

// 중위 순회 함수
void inorderHelp(struct TreeNode* curNode, int k) {
    // base case: 노드가 없거나 k번째 노드를 이미 찾았으면 종료
    if (curNode == NULL || found) {
        return;
    }

    // 왼쪽 서브트리 탐색
    inorderHelp(curNode->left, k);

    // 현재 노드 처리
    count++;
    if (count == k) {
        result = curNode->val;
        found = 1;  // k번째 노드를 찾았음을 표시
        return;
    }

    // 오른쪽 서브트리 탐색
    inorderHelp(curNode->right, k);
}

// k번째로 작은 노드 찾기 함수
int kthSmallest(struct TreeNode* root, int k) {
    // 전역 변수 초기화
    count = 0;
    result = 0;
    found = 0;

    // 중위 순회 시작
    inorderHelp(root, k);

    // 결과 반환
    return result;
}

// TreeNode를 생성하는 헬퍼 함수 (필요하면 사용)
struct TreeNode* createNode(int val) {
    struct TreeNode* node = (struct TreeNode*)malloc(sizeof(struct TreeNode));
    node->val = val;
    node->left = NULL;
    node->right = NULL;
    return node;
}

// 테스트 예제
int main() {
    // 트리 생성
    struct TreeNode* root = createNode(3);
    root->left = createNode(1);
    root->right = createNode(4);
    root->left->right = createNode(2);

    // k번째로 작은 값 찾기
    int k = 2;
    int result = kthSmallest(root, k);

    // 결과 출력
    printf("The %d-th smallest element is %d\n", k, result);

    return 0;
}

File: Binary_tree_general/leetcode/230_Kth_Smallest_Element_in_a_BST/230_Kth_Smallest_Element_in_a_BST.py
### DFS INORDER

# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution1:
    def kthSmallest(self, root: Optional[TreeNode], k: int) -> int:
        self.count = 0
        self.result = None
        self.inorderHelp(root, k)

        return self.result
        

    def inorderHelp(self, curNode, k) -> int :

        if not curNode or self.result is not None:
            return
        
        self.inorderHelp(curNode.left, k)
        self.count +=1 

        if self.count == k :
            self.result = curNode.val
        
        self.inorderHelp(curNode.right, k)



### Using Stack

# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution2:
    def kthSmallest(self, root: Optional[TreeNode], k: int) -> int:
        stack = []
        curr = root

        while stack or curr:
            while curr:
                stack.append(curr)
                curr = curr.left ## 왼쪽으로 끝까지 내려갔다가
            curr = stack.pop() ## pop을 하면서 
            k -= 1
            if k == 0: return curr.val ## k번째 작은 것 return
            curr = curr.right ## right가 있으면 right로 내려감